components/external_child_process/child_monitor.cc

```diff
@@ -0,0 +1,186 @@
+  flags |= FD_CLOEXEC;
+  if (fcntl(fd, F_SETFD, flags) == -1)
+    VLOG(0) << "set fd flags errno:" << errno;
+}
+#endif
+
+static void SetupPipeHack() {
+#if defined(OS_MAC)
+  if (pipe(pipehack) == -1)
+    VLOG(0) << "pipehack errno:" << errno;
+  SetupFD(pipehack[0]);
+  SetupFD(pipehack[1]);
+#else
+  if (pipe2(pipehack, O_CLOEXEC) == -1)
+    VLOG(0) << "pipehack errno:" << errno;
+#endif
+
+  int flags;
+  if ((flags = fcntl(pipehack[1], F_GETFL)) == -1)
+    VLOG(0) << "get flags errno:" << errno;
+  // Nonblock write end on SIGCHLD handler which will notify monitor thread
+  // by sending one byte to pipe whose read end is blocked and wait for
+  // SIGCHLD to arrives to avoid busy reading
+  flags |= O_NONBLOCK;
+  if (fcntl(pipehack[1], F_SETFL, flags) == -1)
+    VLOG(0) << "set flags errno:" << errno;
+
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIGCHLDHandler;
+  sigaction(SIGCHLD, &action, NULL);
+}
+
+static void TearDownPipeHack() {
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIG_DFL;
+  sigaction(SIGCHLD, &action, NULL);
+  close(pipehack[0]);
+  close(pipehack[1]);
+}
+#endif
+
+void MonitorChild(base::ProcessHandle p_handle,
+                  base::OnceCallback<void(base::ProcessId)> callback) {
+  DCHECK(callback);
+#if defined(OS_POSIX)
+  char buf[PIPE_BUF];
+
+  while (1) {
+    if (read(pipehack[0], buf, sizeof(buf)) > 0) {
+      pid_t pid;
+      int status;
+
+      if ((pid = waitpid(base::GetProcId(p_handle), &status, WNOHANG)) != -1) {
+        if (WIFSIGNALED(status)) {
+          VLOG(0) << "child(" << pid << ") got terminated by signal "
+                  << WTERMSIG(status);
+        } else if (WCOREDUMP(status)) {
+          VLOG(0) << "child(" << pid << ") coredumped";
+        } else if (WIFEXITED(status)) {
+          VLOG(0) << "child(" << pid << ") exit (" << WEXITSTATUS(status)
+                  << ")";
+        }
+        std::move(callback).Run(pid);
+      }
+    } else {
+      // pipes closed
+      break;
+    }
+  }
+#elif defined(OS_WIN)
+  WaitForSingleObject(p_handle, INFINITE);
+  std::move(callback).Run(base::GetProcId(p_handle));
+#else
+#error unsupported platforms
+#endif
+}
+
+}  // namespace
+
+ChildMonitor::ChildMonitor()
+    : child_monitor_thread_(new base::Thread("child_monitor_thread")) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+#if defined(OS_POSIX)
+  SetupPipeHack();
+#endif
+  if (!child_monitor_thread_->Start()) {
+    NOTREACHED();
+  }
+}
+ChildMonitor::~ChildMonitor() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+#if defined(OS_POSIX)
+  TearDownPipeHack();
+#endif
+
+  if (child_process_.IsValid()) {
+    child_process_.Terminate(0, true);
+#if defined(OS_MAC)
+    // TODO(https://crbug.com/806451): The Mac implementation currently blocks
+    // the calling thread for up to two seconds.
+    base::PostTask(
```

Comment:
`base::ThreadPool::PostTask`

Reply:
fixed in 76857501ecbfb2f31ff1023ec197de6d90d62618